programs/cleos/main.cpp

```diff
@@ -995,14 +996,35 @@ void ensure_keosd_running(CLI::App* app) {
         pargs.push_back("--unix-socket-path");
         pargs.push_back(string(key_store_executable_name) + ".sock");
 
+        bp::pstream keos_out;
         ::boost::process::child keos(binPath, pargs,
                                      bp::std_in.close(),
-                                     bp::std_out > bp::null,
+                                     bp::std_out > keos_out,
                                      bp::std_err > bp::null);
         if (keos.running()) {
             std::cerr << binPath << " launched" << std::endl;
-            keos.detach();
-            try_local_port(2000);
+
+            std::string keos_msg;
+            constexpr size_t TOTOAL_RETRIES = 10;
+            constexpr size_t KEOS_WAIT_TIME_MS = 200;
+            size_t tries = 0;
+            for(; tries<TOTOAL_RETRIES; ++tries)
+            {
+               std::this_thread::sleep_for(std::chrono::milliseconds(KEOS_WAIT_TIME_MS));
+               std::getline(keos_out, keos_msg);
```

Comment:
getline is blocking.  There's no need to wait.

Reply:
It behaves as syncing, if keosd returns -1 before (daemonizing) sending "keosd::config::keosd_started_msg",  pstream will make getline return immediately.  Then, cleos knows that retry fails, print out err msg, then move on, no wait.

if keosd is already there before cleos runs, cleos won't wait. Otherwise, cleos starts keosd, and wait for keosd  finishing startup  (right before daemonizing).  For any reason, if keosd hangs before startup is done, cleos will hang there too.  

If cleos should run independently from keosd, then correct, will change to asnc read instead.

Reply:
It behaves as syncing, if keosd returns -1 before (daemonizing) sending "keosd::config::keosd_started_msg", pstream will make getline return immediately. Then, cleos knows that retry fails, print out err msg, then move on, no wait.

if keosd is already there before cleos runs, cleos won't wait. Otherwise, cleos starts keosd, and wait for keosd finishing startup (right before daemonizing). For any reason, if keosd hangs  within startup, cleos will hang there too.

If cleos should run independently from keosd, then correct, will change to asnc read instead.

Reply:
The new approach is making http_plug startup yield to wallet-api startup. So, no change here.